Prohibit updating spec.databaseInstance

kajinamit opened the pull request:
Updating spec.databaseInstance does not migrate existing data to a new instance and can break resource access unexpectedly. This prohibits such update.

gibizer commented:
/lgtm

kajinamit committed:
Prohibit updating spec.databaseInstance

Updating spec.databaseInstance does not migrate existing data to a new
instance and can break resource access unexpectedly. This prohibits
such update.

SeanMooney commented:
/hold i understand why you are doing this but i was literally talking about how we could move form a single shared db to per service dbs if we needed too for scaleing reasons by doing a db dump/restores and updating the db reference.

so before we block this i want to at least discuss if we should

stuggi commented:
Maybe we want to do something similar to what we did in the infra-operator to ipset where we added an immutable [1] parameter. if `true` on the old spec, the webhook [2] blocks updating the CR spec, except of the immutable parameter. This forces an additional step to allow changes and sends a  "be careful" signal. Depending of the service operator/CR this immutable parameter could default to true/false.

[1] https://github.com/openstack-k8s-operators/infra-operator/blob/main/apis/network/v1beta1/ipset_types.go#L48
[2] https://github.com/openstack-k8s-operators/infra-operator/blob/main/apis/network/v1beta1/ipset_webhook.go#L102

kajinamit commented:
> /hold i understand why you are doing this but i was literally talking about how we could move form a single shared db to per service dbs if we needed too for scaleing reasons by doing a db dump/restores and updating the db reference.
> 
> so before we block this i want to at least discuss if we should

I understand that possible scenario to migrate database but if we don't have specific timeline to implement that support we should forbid this update. With current implementations, in case a user updates databaseInstance then it switches the database without data migration and it does not remove the existing database in the old instance properly. I'm more concerned about the potential problems caused by unexpected changes.
Anyway I'll bring this to the sync discussionsa this week.

kajinamit commented:
> Maybe we want to do something similar to what we did in the infra-operator to ipset where we added an immutable [1] parameter. if `true` on the old spec, the webhook [2] blocks updating the CR spec, except of the immutable parameter. This forces an additional step to allow changes and sends a "be careful" signal. Depending of the service operator/CR this immutable parameter could default to true/false.
> 
> [1] https://github.com/openstack-k8s-operators/infra-operator/blob/main/apis/network/v1beta1/ipset_types.go#L48 [2] https://github.com/openstack-k8s-operators/infra-operator/blob/main/apis/network/v1beta1/ipset_webhook.go#L102

That's an option, I think. One thing we probably have to consider here is that we have to control whether a specific key should be immutable or not. We can probably add a dict option like debug and use it.

gibizer commented on api/v1beta1/placementapi_webhook.go:
```diff
@@ -86,7 +92,25 @@ func (r *PlacementAPI) ValidateCreate() error {
 func (r *PlacementAPI) ValidateUpdate(old runtime.Object) error {
```
please note that this needs to be called from openstack-operator to meaningfully reject the update. See https://github.com/openstack-k8s-operators/openstack-operator/pull/355

kajinamit replied:
Yes. I'm aware of that requirement.

kajinamit replied:
I'd work on that integration separately so I'll resolve this conversation now but please let me know if you want me to address anything early within this change.

gibizer commented on test/functional/placementapi_webhook_test.go:
```diff
@@ -0,0 +1,57 @@
+					instance.Spec.DatabaseInstance = "changed"
+					return nil
+				})
+			Expect(err).To(HaveOccurred())
```
Please check that the correct error is occurred. As this test now will pass even if the k8s returns http 500 for example.

kajinamit replied:
Fixed in the latest version

gibizer commented on test/functional/suite_test.go:
```diff
@@ -127,13 +134,26 @@ var _ = BeforeSuite(func() {
+
+	err = (&placementv1beta1.PlacementAPI{}).SetupWebhookWithManager(k8sManager)
+	Expect(err).NotTo(HaveOccurred())
+
```
Please add a check that the webhook is up before start executing tests. See https://github.com/openstack-k8s-operators/nova-operator/blob/fe606377229a26837eb263cf99d3e7eb4d3b449b/test/functional/suite_test.go#L235-L245

kajinamit replied:
Fixed in the latest version

SeanMooney commented:
for what it's worth i like martins suggestion.

we can make this immutable by default after the creation of the CR.
timeline for supporting the data migration is probably not until next year.

I think we likely could support db instance migration before GA but we won't be prioritising this until we start looking at things like updates/upgrades in Q1 or later I think.

we would probably couple this with backup/restore, edge or disaster recovery day two operations.

kajinamit committed:
Add flag to allow unsupported operations

This introduces the AdvancedMode spec which allows users to bypass
validations about unsupported operations. Right now this can be used
to update spec.databaseInstance, which is prohibited because of no
automated data migration.

gibizer commented:
> > Maybe we want to do something similar to what we did in the infra-operator to ipset where we added an immutable [1] parameter. if `true` on the old spec, the webhook [2] blocks updating the CR spec, except of the immutable parameter. This forces an additional step to allow changes and sends a "be careful" signal. Depending of the service operator/CR this immutable parameter could default to true/false.
> > [1] https://github.com/openstack-k8s-operators/infra-operator/blob/main/apis/network/v1beta1/ipset_types.go#L48 [2] https://github.com/openstack-k8s-operators/infra-operator/blob/main/apis/network/v1beta1/ipset_webhook.go#L102
> 
> That's an option, I think. One thing we probably have to consider here is that we have to control whether a specific key should be immutable or not. We can probably add a dict option like debug and use it.

I have a bad feeling about the UX of having a separate map defining what field is settable and when. But I have no good suggestion how to force an extra round-trip for the user to make sure the DB change is intentional.

gibizer commented on api/v1beta1/placementapi_types.go:
```diff
@@ -86,6 +86,11 @@ type PlacementAPISpec struct {
+	// +kubebuilder:validation:Optional
+	// PlacementAPIImmutable - make some specs immutable to prohibit update operations
+	// not fully supported by this operator.
+	Immutable PlacementAPIImmutable `json:"immutable,omitempty"`
```
I have a bad feeling about the UX of having a separate struct / map defining what field is settable and when. But I have no good suggestion how to force an extra round-trip for the user to make sure the DB change is intentional.

Maybe(?) we can have a single "advanced-mode" flag for the whole CRD. That is defaulted to false, and prevents any dangerous operation. That would allow us not to duplicate fields directly.

kajinamit replied:
This is addressed in the latest version.

gibizer commented on api/v1beta1/placementapi_types.go:
```diff
@@ -86,6 +86,11 @@ type PlacementAPISpec struct {
 	Debug PlacementAPIDebug `json:"debug,omitempty"`
 
+	// +kubebuilder:validation:Optional
+	// PlacementAPIImmutable - make some specs immutable to prohibit update operations
```
Please describe the logic at the DatabaseInstance field too.

kajinamit replied:
I've added the description of that field here: https://github.com/openstack-k8s-operators/placement-operator/pull/17/files#diff-e5c063113f25a841c51f8370d6711ece1751aa814b24f925f440881290cb6e59R170

Or do you intend to adding any additional description to this immutable key level ?

kajinamit replied:
This is no longer applicable in the latest version.

gibizer commented on api/v1beta1/placementapi_types.go:
```diff
@@ -159,6 +164,14 @@ type PasswordSelector struct {
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=true
+	// Make databaseInstance immutable
+	DatabaseInstance bool `json:"databaseInstance"`
```
This will not work well with webhooks and json ser/deserialization. The webhook ser/des step will default this to false (the golang default) when the field is not provided in the request. You need to make this a *bool to avoid this.

kajinamit replied:
Fixed in the latest version.

gibizer commented on api/v1beta1/placementapi_webhook.go:
```diff
@@ -86,7 +92,27 @@ func (r *PlacementAPI) ValidateCreate() error {
+					Resource: GroupVersion.WithKind("PlacementAPI").Kind,
+				}, r.GetName(), &field.Error{
+					Type:     field.ErrorTypeForbidden,
+					Field:    "*",
```
I think we should report the which field we are reporting the error on here. E.g. this should be "spec.databaseInstance" if the local webhook validated it, or "spec.placement.template.databaseInstance" if the openstack-operator validated it. You can pass a field.Path here to get the necessary base path of the field for the two different call path. See https://github.com/openstack-k8s-operators/nova-operator/pull/411/files#diff-0bb4bc5fee241020d547ea9eb775efb27b4d8082e09ad7d51a221cebb8cd6002R110 for example

kajinamit replied:
Fixed in the latest version

gibizer commented on test/functional/placementapi_webhook_test.go:
```diff
@@ -0,0 +1,72 @@
+				})
+			Expect(err).NotTo(HaveOccurred())
+		})
+	})
```
Please add a test case that shows that defaulting for the boolean field works as expected

kajinamit replied:
I've added the test case for default.

stuggi commented:
> > Maybe we want to do something similar to what we did in the infra-operator to ipset where we added an immutable [1] parameter. if `true` on the old spec, the webhook [2] blocks updating the CR spec, except of the immutable parameter. This forces an additional step to allow changes and sends a "be careful" signal. Depending of the service operator/CR this immutable parameter could default to true/false.
> > [1] https://github.com/openstack-k8s-operators/infra-operator/blob/main/apis/network/v1beta1/ipset_types.go#L48 [2] https://github.com/openstack-k8s-operators/infra-operator/blob/main/apis/network/v1beta1/ipset_webhook.go#L102
> 
> That's an option, I think. One thing we probably have to consider here is that we have to control whether a specific key should be immutable or not. We can probably add a dict option like debug and use it.

do we have to be that specific on a per field via the CRD? why not just have a single `immutable` parameter. if that set all fields which should not be possible to change get blocked by the webhook. if for some reason one of those need to be changed the `immutable` param needs to set to false. in that case you could change any of them, but you have to do that additional step.

gibizer commented:
> > > Maybe we want to do something similar to what we did in the infra-operator to ipset where we added an immutable [1] parameter. if `true` on the old spec, the webhook [2] blocks updating the CR spec, except of the immutable parameter. This forces an additional step to allow changes and sends a "be careful" signal. Depending of the service operator/CR this immutable parameter could default to true/false.
> > > [1] https://github.com/openstack-k8s-operators/infra-operator/blob/main/apis/network/v1beta1/ipset_types.go#L48 [2] https://github.com/openstack-k8s-operators/infra-operator/blob/main/apis/network/v1beta1/ipset_webhook.go#L102
> > 
> > 
> > That's an option, I think. One thing we probably have to consider here is that we have to control whether a specific key should be immutable or not. We can probably add a dict option like debug and use it.
> 
> do we have to be that specific on a per field via the CRD? why not just have a single `immutable` parameter. if that set all fields which should not be possible to change get blocked by the webhook. if for some reason one of those need to be changed the `immutable` param needs to set to false. in that case you could change any of them, but you have to do that additional step.

Yeah, that feels a bit easier UX. I said above:
> Maybe(?) we can have a single "advanced-mode" flag for the whole CRD. That is defaulted to false, and prevents any dangerous operation. That would allow us not to duplicate fields directly.

stuggi commented:
> > > > Maybe we want to do something similar to what we did in the infra-operator to ipset where we added an immutable [1] parameter. if `true` on the old spec, the webhook [2] blocks updating the CR spec, except of the immutable parameter. This forces an additional step to allow changes and sends a "be careful" signal. Depending of the service operator/CR this immutable parameter could default to true/false.
> > > > [1] https://github.com/openstack-k8s-operators/infra-operator/blob/main/apis/network/v1beta1/ipset_types.go#L48 [2] https://github.com/openstack-k8s-operators/infra-operator/blob/main/apis/network/v1beta1/ipset_webhook.go#L102
> > > 
> > > 
> > > That's an option, I think. One thing we probably have to consider here is that we have to control whether a specific key should be immutable or not. We can probably add a dict option like debug and use it.
> > 
> > 
> > do we have to be that specific on a per field via the CRD? why not just have a single `immutable` parameter. if that set all fields which should not be possible to change get blocked by the webhook. if for some reason one of those need to be changed the `immutable` param needs to set to false. in that case you could change any of them, but you have to do that additional step.
> 
> Yeah, that feels a bit easier UX. I said above:
> 
> > Maybe(?) we can have a single "advanced-mode" flag for the whole CRD. That is defaulted to false, and prevents any dangerous operation. That would allow us not to duplicate fields directly.

right, could not connect the "advanced-mode" word with immutable on quick scanning your updates. we should just have the same param name across the operators. we used `Immutable`  https://github.com/openstack-k8s-operators/infra-operator/blob/main/apis/network/v1beta1/ipset_types.go#L46-L48  , but we only just introduced it, so can be changed if we agree on something else. I just realized, I missed a description there ...

kajinamit commented:
OK I agree that we can make it a single parameter. The reason why I preferred flags per parameters is that I wanted to make the flag name specific enough to help users understand the expectation.

Probably we want to name it `disableValidation` instead of `advancedMode`. The implication of advanced mode is quite vague and I'd prefer using more specific name, unless we have any other use cases we want to enable by that flag.

gibizer commented:
> > > > > Maybe we want to do something similar to what we did in the infra-operator to ipset where we added an immutable [1] parameter. if `true` on the old spec, the webhook [2] blocks updating the CR spec, except of the immutable parameter. This forces an additional step to allow changes and sends a "be careful" signal. Depending of the service operator/CR this immutable parameter could default to true/false.
> > > > > [1] https://github.com/openstack-k8s-operators/infra-operator/blob/main/apis/network/v1beta1/ipset_types.go#L48 [2] https://github.com/openstack-k8s-operators/infra-operator/blob/main/apis/network/v1beta1/ipset_webhook.go#L102
> > > > 
> > > > 
> > > > That's an option, I think. One thing we probably have to consider here is that we have to control whether a specific key should be immutable or not. We can probably add a dict option like debug and use it.
> > > 
> > > 
> > > do we have to be that specific on a per field via the CRD? why not just have a single `immutable` parameter. if that set all fields which should not be possible to change get blocked by the webhook. if for some reason one of those need to be changed the `immutable` param needs to set to false. in that case you could change any of them, but you have to do that additional step.
> > 
> > 
> > Yeah, that feels a bit easier UX. I said above:
> > > Maybe(?) we can have a single "advanced-mode" flag for the whole CRD. That is defaulted to false, and prevents any dangerous operation. That would allow us not to duplicate fields directly.
> 
> right, could not connect the "advanced-mode" word with immutable on quick scanning your updates. we should just have the same param name across the operators. we used `Immutable` https://github.com/openstack-k8s-operators/infra-operator/blob/main/apis/network/v1beta1/ipset_types.go#L46-L48 , but we only just introduced it, so can be changed if we agree on something else. I just realized, I missed a description there ...

I guess the naming should reflect the effect of the flag. If we only control immutability with it then sure we should call it Immutable. But I'm wondering that there might be other situations where we want to disable complex validations. E.g. we have a validation webhook check in nova that you should not deploy metadata in cell0. So disabling such complex validation would not be an immutability change but more like an advanced operation that we only allow if you are super sure about it. (Note that the nova example might be a wrong one as I don't see why somebody want to ever deploy metadata on cell0, but similar complex rule might exists elsewhere).

gibizer commented:
> OK I agree that we can make it a single parameter. The reason why I preferred flags per parameters is that I wanted to make the flag name specific enough to help users understand the expectation.
> 
> Probably we want to name it `disableValidation` instead of `advancedMode`. The implication of advanced mode is quite vague and I'd prefer using more specific name, unless we have any other use cases we want to enable by that flag.

With `disableValidation` I have another problem (sorry it seems I have that guy today who does not like anything). We have validations we don't want to disable with this check. E.g. there is a cell name length validation that cannot be disabled as long cell names will always lead to CR creation error in nova-operator (as we are concatenating names). So in this case the vague naming might help to force us defining in the description of the field the exact meaning. I.e. which specific checks are disabled by that flag.

stuggi commented:
> > OK I agree that we can make it a single parameter. The reason why I preferred flags per parameters is that I wanted to make the flag name specific enough to help users understand the expectation.
> > Probably we want to name it `disableValidation` instead of `advancedMode`. The implication of advanced mode is quite vague and I'd prefer using more specific name, unless we have any other use cases we want to enable by that flag.
> 
> With `disableValidation` I have another problem (sorry it seems I have that guy today who does not like anything). We have validations we don't want to disable with this check. E.g. there is a cell name length validation that cannot be disabled as long cell names will always lead to CR creation error in nova-operator (as we are concatenating names). So in this case the vague naming might help to force us defining in the description of the field the exact meaning. I.e. which specific checks are disabled by that flag.

agreed, we should not skip all validations, correctness needs still to be checked.

SeanMooney commented:
i dont partcallay like disableValidation as we are not going to disabel all validation.
im -1 on having a flag to disable all validation. this is also not really a validation

a validation would simply be checking that the db instance exists and possibly that the table exists. 

maintenance mode or similar i think would be better capture that you are only performing a day two operation such as relocating the db. in mantance mode we could do something like scale down to 0 API instance until you set it to false so  the workflow would be:

1. enable maintance mode (scale down placement to prevent new requests)
2. dump the db
3. restore it to and different db server
4. update the reference to the new instance and unset maintance mode

in 1 we would scale down the API so that no new requests can happen 
2/3 would eventually be automated 
4 would restore api access with the container pointing at the new db.

any attempt to update the db outside of maintance mode would be blocked.

SeanMooney commented:
just to be clear if we add a maintenance mode option I'm ok with relaxing validation slightly during mantaince mode as long as we reseert it when exiting it.

gibizer commented:
I can compromise on "maintenanceMode" over "advancedMode" or "sureIKnowWhatI'mDoingMode" (OK, I'm kidding on the latter), but "maintenanceMode" still feels a bit too specific to one given process.

gibizer commented:
> just to be clear if we add a maintenance mode option I'm ok with relaxing validation slightly during mantaince mode as long as we reseert it when exiting it.

Ohh that is a very good point. We need to be careful if this mode allows persisting a value that otherwise would be rejected not because of the fact that it is a change but because of the specific new value itself, then we cannot re-enable the strict mode any more.

kajinamit commented:
I'd still disagree with the raw `maintenanceMode` because the aim of `maintenance` is not clear. We should probably name it `databaseMaintenanceMode` though that does not make much difference from the original proposal to have specific flag to make databaseInstance mutable ...

bshephar commented:
Just adding my 2 cents in here. But I feel like if we _need_ a maintenance mode, we might be making things a bit too complicated. Like you just discussed in the later comments, it introduces more complexity that we need to handle somewhere.

If we wanted some kind of migration facility, wouldn't it make more sense to provide some interface for that during the `reconcileUpdate` workflow? Since we expect to be handling schema updates and such during that update workflow, adding some migration workflow there would probably be easier? Rather than having some bespoke maintenance mode?

Also, does seperate DB's for each service mean we're going to run a mysql pod per service? That feels a bit heavy, and I'm not sure I understand what we're going to gain at the scale 90% of our users are at. Maybe in the like the top 5% of large deployments there would be some performance benefit.

stuggi commented:
> I'd still disagree with the raw `maintenanceMode` because the aim of `maintenance` is not clear. We should probably name it `databaseMaintenanceMode` though that does not make much difference from the original proposal to have specific flag to make databaseInstance mutable ...

the thing is will it only be the database, or is some other parameter coming next. Placement is simple and does not have a lot of params, but for other services there are probably more parameters which per default should not change after the initial create.

kajinamit commented:
Updated the PR to clear some of the earlier comments. We still have to agree with the overall design but I kind of feel we are unsure about the plan to support any day 2 operators. If we take the position to "avoid prohibiting anything we can later allow" then I'd probably abandon the whole my these validations works, just hoping users would not do anything unsupported.

bshephar commented:
> Updated the PR to clear some of the earlier comments. We still have to agree with the overall design but I kind of feel we are unsure about the plan to support any day 2 operators. If we take the position to "avoid prohibiting anything we can later allow" then I'd probably abandon the whole my these validations works, just hoping users would not do anything unsupported.

I'm just concerned that we will have users accidentally changing this and ending up in a mess. We need to consider the type of support cases that are going to come up, and what the actual recovery procedure will look like. My concern would be that if this particular issue come up, then it will automatically become an escalation and the recovery of it would be messy.

SeanMooney commented:
my original thought was in the future we would just automatically copy the db if you updated the DB instance for what its worth

I didn't think we would need to have any flag to enable that.

the reason i suggested maintenanceMode and not something like databaseMaintenanceMode
is I don't want to see a pattern of <funcaltiy>MantianceMode propagated in the operators

so if we had a flag at all I wanted to have one that was very generic.


I'm still not really sold on the idea that we need this to be imuatbale and would prefer to trust our users to understand that if the change the database reference it will print to a different db.

kajinamit commented:
> the reason i suggested maintenanceMode and not something like databaseMaintenanceMode is I don't want to see a pattern of MantianceMode propagated in the operators
> so if we had a flag at all I wanted to have one that was very generic.

I understand your points but this requires that we consider all of the `maintenance` tasks we support in all services, so that we understand what should be allowed in maintenance mode, which would too much expand scope of this work.

> would prefer to trust our users to understand that if the change the database reference it will print to a different db.

While maintaining TripleO, we've seen several unfortunate cases where deployments were messed up by wrong edit in  templates. Users might understand the impact of each edit(Note even some users had no idea about the changes they were making in their production, for real) but preventing any unintentional edit is merely impossible. Imagine the case where user do `oc edit openstackcontrolplane/openstack` and unintentionally modified `databasInstance: openstack` to `databaseInstance: openstac` while editing the spec by editor. I believe this can be quite possible scenario we can face.

kajinamit commented:
> Imagine the case where user do oc edit `openstackcontrolplane/openstack` and unintentionally modified `databasInstance: openstack` to `databaseInstance: openstac` while editing the spec by editor. 

Giving it another thought then this might not cause much serious result as it fails at an earlier point unless there is a DB instance with `openstack` deployed ... So we don't probably have to consider unintentional edit really.

Though I still feel guarding it is very beneficial for users as well as us. You know how many users changed AZ of existing compute nodes after they created instance on these hosts. Such "unsupported" operation can't be really blocked unless we prohibit these strictly at validation.

SeanMooney commented:
> changed AZ of existing compute nodes after they created instance on these hosts.

as an aside that specific example is a bug in the nova API i know how to fix that but we just have not yet
we missed blocking it in one specific edgecase. we do block that in most case except one specific case.

bogdando commented:
Adding immutability control flags look overcomplicating the API (CRD) design. There are better examples to follow, like immutability control over the [ReplicaSet specs](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#label-selector-updates). The controller/validation logic should be taking care of it, not asked to be controlled by users.

Neither we need ``disableValidation`` flags in CRD, becase for that, the way to go is deploy w/o webhooks

stuggi commented:
> Adding immutability control flags look overcomplicating the API (CRD) design. There are better examples to follow, like immutability control over the [ReplicaSet specs](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#label-selector-updates). The controller/validation logic should be taking care of it, not asked to be controlled by users.
> 
> Neither we need `disableValidation` flags in CRD, becase for that, the way to go is deploy w/o webhooks

not sure if I get this correct, but this won't work in our case. this would mean if there is a required change for any immutable parameter you'd have to recreate (delete/create) it, or do I get your point wrong?

kajinamit commented:
Giving it another thought after discussion with @stuggi and drafting some parameter descriptions, I now think the `advancedMode` is the most appropriate naming for the flag (among the ones proposed so far), because the flag basically allows usage not supported by the operator . Alternative option I can think of was `unsafe` or `unsupported` but these can imply different meanings.  Also I'd prefer advanced to maintenance because maintenance is too much vague.

stuggi commented on api/v1beta1/placementapi_webhook.go:
```diff
@@ -86,7 +92,25 @@ func (r *PlacementAPI) ValidateCreate() error {
+		return apierrors.NewInternalError(fmt.Errorf("unable to convert existing object"))
+	}
+
+	if ! r.Spec.AdvancedMode {
```
do we want an explicite extra step to toggle the `advancedMode` ? then we should check this in the old object. if we are ok that updating `advancedMode` and other parameters in one go this is fine.

gibizer commented on api/v1beta1/placementapi_types.go:
```diff
@@ -46,9 +46,9 @@ type PlacementAPISpec struct {
-	// Right now required by the maridb-operator to get the credentials from the instance to create the DB
-	// Might not be required in future
+	// MariaDB instance name.
+	// This can't be updated unless advanced mode is enabled, because updating this does not migrate
+	// data to a new database instance.
```
Will updating this to a new database instance at least ensures that the DB schema is applied to the new DB instance?

gibizer commented on api/v1beta1/placementapi_webhook.go:
```diff
@@ -86,7 +92,25 @@ func (r *PlacementAPI) ValidateCreate() error {
+				),
+			)
+		}
+	}
```
This logic needs to be in a function that is callable from openstack-operator validation webhook. The pattern is documented in https://github.com/openstack-k8s-operators/docs/blob/main/webhooks.md#reusing-webhook-logic

gibizer commented on api/v1beta1/placementapi_webhook.go:
```diff
@@ -86,7 +92,25 @@ func (r *PlacementAPI) ValidateCreate() error {
+					Resource: GroupVersion.WithKind("PlacementAPI").Kind,
+				}, r.GetName(), field.Forbidden(
+					field.NewPath("spec").Child("databaseInstance"),
+					"Value is immutable.",
```
I would mention advancedMode here.

gibizer commented on api/v1beta1/placementapi_types.go:
```diff
@@ -86,6 +86,11 @@ type PlacementAPISpec struct {
 
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	// AdvancedMode - allows some usage not fully supported by this operator.
```
We need to be more explicit about documenting what is allowed.

bshephar replied:
> We need to be more explicit about documenting what is allowed.

For me, this `advancedMode` terminology is too broad for this reason. We need to assume that at best users might read the documentation, but most likely they will just read the spec. And `advancedMode` doesn't tell me anything about what it does. Is this some kind of Placement advancedMode? Can I find out about it by reading the Placement docs? Does it allow the operator do advanced things? 

I personally would prefer to see something like feature sets, where we enable and disable specific features with clearly defined scopes. eg. `spec.enabledFeatures.databaseMigration`. This to me would be a more intuitive interface, and the expectations of such a setting are fairly obvious from the naming convention.

My other question would be, how common of an interface is this in other Kubernetes primitives or operators? Do we see boolean configurable "modes" elsewhere? Or is our implementation bespoke and therefore not quite as  intuitive to the average Kubernetes user? Keeping in mind that a large selling point for this solution is "You don't need to have TripleO specific knowledge / Heat specific knowledge / Red Hat specific knowledge. You just need to know kubernetes." I think from a design perspective, we should keep that fairly high on the priority list of considerations.

gibizer replied:
> For me, this advancedMode terminology is too broad for this reason. We need to assume that at best users might read the documentation, but most likely they will just read the spec. And advancedMode doesn't tell me anything about what it does. Is this some kind of Placement advancedMode? Can I find out about it by reading the Placement docs? Does it allow the operator do advanced things?

It is the latter. :) 

On this point I'm OK to group the operator specific configuration if there are many (like advancedMode, preserveJobs) into a separate struct to express that they are distinct from the configuration that is for the deployed OpenStack service (like customeServiceConfig). 

> I personally would prefer to see something like feature sets, where we enable and disable specific features with clearly defined scopes. eg. spec.enabledFeatures.databaseMigration. This to me would be a more intuitive interface, and the expectations of such a setting are fairly obvious from the naming convention.

My problem with this that we might end up with a long list of flags if we start adding a configurable for each different "advanced feature". Like the next would be allowing changing the memcached instance under a flag. 

> My other question would be, how common of an interface is this in other Kubernetes primitives or operators? Do we see boolean configurable "modes" elsewhere? Or is our implementation bespoke and therefore not quite as intuitive to the average Kubernetes user? 

I'm not sure how we define 'boolean configurable "modes"' but it was fairly easy to find examples of boolean flags in Spec structs that's value affecting the behavior of the operator itself instead of affecting the behavior of the running service deployed by the operator:
* https://github.com/openstack-k8s-operators/rabbitmq-cluster-operator/blob/f8d9ac5a263c5d9c82dfeef8170eacf16c1cc5e1/api/v1beta1/rabbitmqcluster_types.go#L76-L80 
* JobSpec.Suspended https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#job-v1-batch 

> Keeping in mind that a large selling point for this solution is "You don't need to have TripleO specific knowledge / Heat specific knowledge / Red Hat specific knowledge. You just need to know kubernetes." I think from a design perspective, we should keep that fairly high on the priority list of considerations.

If we allow everything that k8s allows and ignoring what OpenStack supports from the set of operations k8s allows, then I think we will end up in too many dangerous possibilities. I think one of the value we can bring to the customers is codified knowledge about deploying OpenStack, encoded into the implementation of the operators. If we know that changing a DB server under a running OpenStack deployment is dangerous for the integrity of the deployment, then we should tell that to our users somehow. I also agree that we should not assume that they read the doc. So I would codify that knowledge into the operator.

bshephar replied:
Hey @gibizer thanks for the response.

I think in both of those examples it's more clear what the setting is though. It's not a broadly scoped `advancedMode` is more my point there. I think we want to avoid a situation where people just think, "Advanced mode! I want to do advanced things! I'll set this to true."

The alternative of `databaseMigration` makes you think a little more. "Do I want to be doing a database migration? I don't think so. I'll leave it false."

It's not that we want to allow k8s things. We just want to make sure the interface is intuitive for people familiar with k8s. I'm not too sure there are examples of other operators implementing an `advancedMode` that allows an arbitrary list of operations is all. In this case, we're talking pretty specifically about a k8s operator feature where we would implement job logic to migrate databases around. The OpenStack service doesn't necessarily need to be aware or care, it just needs a new address for the DB service and a restart. So I think in this case, we're implementing an interface on a k8s CRD which seems more specific to a k8s implementation detail rather than OpenStack implementation detail.

But yeah, I understand your concern around an all encompassing list of possible things that can be done. I would guess at a max of 5 per service though which doesn't seem too bad.

In any case, I'm just a single voice. If we have a consensus on the approach proposed here, I'm happy to work with it. But I thought it would be worth at least raising my concern and suggestion for consideration before we merge it.

:)

Thanks again for the perspective and happy Monday.

gibizer replied:
Hi @bshephar.
I have to admit your reasoning make sense and started to convince me to move away from advancedMode. If we can limit ourselves to a reasonable amount of  feature flags, and move them into a separate struct, like you proposed above with `spec.enabledFeatures` then I think I'm on board of this idea.
